### PR TITLE
Removing all sigma spends referencing the current block hash when the…

### PR DESCRIPTION
**Pull request by martun**
… block is removed.

### CODE REVIEW

File: src/zerocoin_v3.cpp [hunk 1]
```diff
@@ -22,6 +23,8 @@
 
 #include <ios>
 
+extern CTxMemPool mempool;
+
```
Comment (ultimaweapon) [inline src/zerocoin_v3.cpp]: I think we don't need this due to we already included `txmempool.h`

File: src/zerocoin_v3.cpp [hunk 2]
```diff
@@ -460,8 +463,41 @@ bool CheckSigmaTransaction(
     return true;
 }
 
-void DisconnectTipSigma(CBlock & /*block*/, CBlockIndex *pindexDelete) {
+void RemoveSigmaSpendsReferencingBlock(CBlockIndex* blockIndex) {
+    std:vector<CTransaction> txn_to_remove;
+    for (CTxMemPool::txiter mi = mempool.mapTx.begin(); mi != mempool.mapTx.end(); ++mi) {
```
Comment (ultimaweapon) [inline src/zerocoin_v3.cpp]: We need to lock before accessing `mempool`

Comment (levonpetrosyan93) [reply src/zerocoin_v3.cpp]: We are locking upper, for example in DisconnectBlocks() . Why do we need to lock here? May be better to call AssertLockHeld() here?

Comment (psolstice) [reply src/zerocoin_v3.cpp]: theoretically it's ok to have it run only under `cs_main` lock which is acquired before the call. But adding one more lock won't hurt

Comment (psolstice) [reply src/zerocoin_v3.cpp]: also have a minor question to @martun : what about dandelion pool? don't we need to delete txs from there as well?

Comment (martun) [reply src/zerocoin_v3.cpp]: Having them left in stempool does not hurt, but I will change the code to remove also from there. Will also add additional locks, they will not hurt.

**Pull request merged**